### DISCUSSION TITLE
Questions on how to use UBA

### DISCUSSION THREAD

**Discussion by otseng** [eliranwong/UniqueBible#130] [Q&A]
Do you have a question on how to use a feature of UBA?  Post it here.

**Comment by onegladiator4u** [comment]
Hi bro,
How can I create and send you guys a debug report?

**Reply by otseng** [reply]
https://github.com/eliranwong/UniqueBible/wiki/FAQ#how-can-i-report-a-bugissue-in-the-uniquebibleapp

**Comment by onegladiator4u** [comment]
I tried python debug.py
It showed like this 
![image](https://user-images.githubusercontent.com/48617390/108502892-86cd4000-72d9-11eb-9bd3-8af4211fce20.png)

**Reply by otseng** [reply]
This will get fixed in the next update.

**Reply by onegladiator4u** [reply]
Ok brother

**Reply by onegladiator4u** [reply]
Brother this didn't fix yet. I'm still getting 
![image](https://user-images.githubusercontent.com/48617390/108815940-4de3e280-75db-11eb-8fa8-39392e82fd40.png)

**Reply by otseng** [reply]
I've created a PR to address this:
https://github.com/eliranwong/UniqueBible/pull/181

Hopefully it'll work after this.

**Reply by onegladiator4u** [reply]
I'll let you know after the update or should I just replace the code in debug file and try ?

**Reply by onegladiator4u** [reply]
It's working now.

**Comment by onegladiator4u** [comment]
Hi @otseng is there a way to set a particular font to a bible version? 
If I set a font to view my native language Bible well, all the other bibles get affected with that.
Can I set set any any value in bible database so that I can use the font for that bible? Or any other way?

**Reply by eliranwong** [reply]
@onegladiator4u  you can achieve it with custom css feature (https://github.com/eliranwong/UniqueBible/wiki/custom_css)

For example, in Chinese bible, CUV, and CUVs, I use a pair of tags to wrap the bible text \<zh\> text text text \</zh\>

and in css I put:

zh { font-family:'myFont'; }

Chinese bibles are then displayed with "myFont"

**Reply by otseng** [reply]
This would be a good feature to have and be automatic without having to create custom code.  Go ahead and create an issue about this.

**Reply by onegladiator4u** [reply]
> 
> 
> @onegladiator4u you can achieve it with custom css feature (https://github.com/eliranwong/UniqueBible/wiki/custom_css)
> 
> For example, in Chinese bible, CUV, and CUVs, I use a pair of tags to wrap the bible text <zh> text text text </zh>
> 
> and in css I put:
> 
> zh { font-family:'myFont'; }
> 
> Chinese bibles are then displayed with "myFont"

I would try this. Thanks

**Reply by onegladiator4u** [reply]
Bro @eliranwong how can I wrap the text with that kind of tags in the Bible ?
That looks very difficult to add tags to each verse one by one. Is there an easy way to add it to both Verse and Bible tables in database?

**Reply by onegladiator4u** [reply]
> This would be a good feature to have and be automatic without having to create custom code. Go ahead and create an issue about this.

Dude, honestly I don't know what to write for this issue. 😀

**Reply by otseng** [reply]
Created.

https://github.com/eliranwong/UniqueBible/issues/187

**Reply by onegladiator4u** [reply]
Thanks man

**Comment by onegladiator4u** [comment]
Hi @otseng ,
I'm working on Malayalam language file. After clicking OK in edit language file, it's not saving. Normally after the restart it would show up the changes. But it's not changing in edit language file window.
Edit working translation is opening up in English. What should I do to work Malayalam in that?

**Reply by otseng** [reply]
Created PR to fix this:
https://github.com/eliranwong/UniqueBible/pull/224

**Reply by onegladiator4u** [reply]
Thank you

**Reply by onegladiator4u** [reply]
May I know what does 'Edit Working Translation' option do?

**Reply by eliranwong** [reply]
@onegladiator4u  I wrote a wiki page for you at: https://github.com/eliranwong/UniqueBible/wiki/Edit-Working-Translation

you may set in config.py:

workingTranslation = 'ml'

so it opens your working translation directly.

**Reply by eliranwong** [reply]
@onegladiator4u 

from version 22.07:

* Now automatically save option of language file as config.workingTranslation when user select a language from dropdown menu displayed with "Edit language file".

You you do not need to edit config.py file on workingTranslation manually.

**Reply by onegladiator4u** [reply]
Thanks guys. That is very helpful.

**Reply by eliranwong** [reply]
@onegladiator4u

also added:

Major change in version 22.08:
* Added "Select Tooltip Translation" in developer menu.

This one saves referenceTranslation value.

**Reply by onegladiator4u** [reply]
@eliranwong bro are you adding new options like "Select Tooltip Translation" to language files? I couldn't find it when I checked.

**Reply by onegladiator4u** [reply]
Should I create issues for spelling mistakes and possible tooltip errors or can I mention it here? @eliranwong

**Reply by eliranwong** [reply]
@onegladiator4u I haven't added a new item in language file.  reason: you are editing one of the language files.

If I add a new item and upload to github, when you update the language files there is a risk that your working file is being overwritten according to current language file settings.  do you do regular backup of the translation you are working on?

@otseng if users edit directly on the language file and they don't use git pull to update, there is a risk their current work will be overwritten.

**Reply by eliranwong** [reply]
> Should I create issues for spelling mistakes and possible tooltip errors or can I mention it here? @eliranwong

@onegladiator4u 
yes, please create issues.  thanks for that.

**Reply by eliranwong** [reply]
@onegladiator4u Please be aware that in version 22.16, language files are updated.  Please make sure you have a backup copy of your working file.

**Reply by otseng** [reply]
@onegladiator4u I would suggest doing all your app updates via git instead of using the UBA updater.  Git would check if there are any files you have modified and prevent overwriting them.  

If not able to use git, then you should save your file prior to letting UBA update it.  You can also periodically send us your translated file so we can incorporate it into an update.

**Reply by onegladiator4u** [reply]
@otseng @eliranwong Now onward I'll backup language files before the update. What all the files should I backup? One file is `language_ml.py`. There is another file named `language_ml.cpython-39.pyc` in `lang\__pycache__` folder. Should I backup that too?

**Reply by eliranwong** [reply]
@onegladiator4u  only language_ml.py is enough.  how many did you translate?  I am working on Chinese translation.  It takes times to translate.

**Reply by onegladiator4u** [reply]
@eliranwong 
> only language_ml.py

Ok.
I don't know know how much I did. Probably 30%. I checked the file. It has more than 600 lines. That's a lot. It'll take some days I guess.

**Comment by onegladiator4u** [comment]
@eliranwong / @otseng May I know what this does? I don't understand what to translate from this. Thanks in advance 
![image](https://user-images.githubusercontent.com/48617390/110015385-cec47c00-7d49-11eb-8dbe-a23b474ebf7a.png)

**Reply by otseng** [reply]
It's Greek for "to execute speedily"

https://www.billmounce.com/greek-dictionary/syntemno

**Reply by eliranwong** [reply]
It is one of the shortcut sets

![Screenshot 2021-03-04 at 7 47 39 PM](https://user-images.githubusercontent.com/25262722/110021447-a412fd00-7d22-11eb-91ee-f95c35b676a9.png)

**Reply by onegladiator4u** [reply]
Thanks guys.

**Comment by onegladiator4u** [comment]
Hi brothers,
I'm currently using 'dark_lightgreen' theme. How can I change the color of verse references ?
I tried to modify 'dark_light green.xml' file. It didn't change the color of it.
![image](https://user-images.githubusercontent.com/48617390/111916776-9ee3db00-8aa2-11eb-9aec-e1d9856ef3ed.png)
Should I add any other value to the xml file or is it in css file? Is that possible ?

**Reply by eliranwong** [reply]
@onegladiator4u  
you can easily do this with custom css feature.  you may read https://github.com/eliranwong/UniqueBible/wiki/custom_css

the value you enter in custom css file will overwrite the original one.

**Reply by eliranwong** [reply]
for example, put the following line in file htmlResources > css > custom.css

> ref { color: blue; }

You will find all reference links turn blue

**Reply by onegladiator4u** [reply]
> ref { color: blue; }

Thank you. I didn't know what name it's used in css.

**Reply by eliranwong** [reply]
@onegladiator4u next time you may check the name by right-click > Copy > HTML

and paste in a text editor and check

**Comment by onegladiator4u** [comment]
Hello guys,
I'm trying to adjust mini control width in config.py. I couldn't. Only height is getting affected when restart.
![image](https://user-images.githubusercontent.com/48617390/115985179-4b483e00-a5c8-11eb-8d54-542fce7c53ae.png)
Numbers are not readable for me. When I adjust by dragging width it's readable .
Is it because of qt theme?

**Reply by eliranwong** [reply]
I think window width can be set to larger value, so don't need to resize manually.

**Reply by onegladiator4u** [reply]
When I make width larger manually it still unreadable. So setting larger value to width may not work. It's readable when I resize it to small. It's like a mask is applied to button texts when resizing.

**Comment by onegladiator4u** [comment]
Does UBA have inbuilt browser ? i remember seeing something like that .

**Reply by otseng** [reply]
online:::https://uniquebible.app

**Reply by onegladiator4u** [reply]
hi osteng .this command opens the external browser . i was thinking about inbuilt one. does uba has it?
i was looking to open site inside the app so that the verse references in the site can be seen in bottom window .

**Comment by eliranwong** [comment]
Also make sure the useWebbrowser config flag is off

**Reply by onegladiator4u** [reply]
oh i missed it. thanks brother

**Comment by eliranwong** [comment]
Also make sure the useWebbrowser config flag is off

**Comment by onegladiator4u** [comment]
Hello brothers, There was a drop down box to select bible version in top . is that gone or is it available through any setting ?

**Comment by otseng** [comment]
It's selectable by the Reference Button Action menu selection.  Select "Direct" to have the drop down.  If "minicontrol", it'll be selectable by the miniControl.  

<img width="646" alt="Screen Shot 2021-07-02 at 2 56 18 PM" src="https://user-images.githubusercontent.com/806056/124317926-ea982d00-db45-11eb-9955-76433e5e93d5.png">

<img width="280" alt="Screen Shot 2021-07-02 at 2 56 33 PM" src="https://user-images.githubusercontent.com/806056/124317942-eec44a80-db45-11eb-8640-467917089f08.png">

**Reply by onegladiator4u** [reply]
Thank you . it's better now. we can access mini-control :).
do you guys have any plans on creating a separate window for settings like. preferences in other softwares.
with tabs for each category and all?

**Reply by otseng** [reply]
> do you guys have any plans on creating a separate window for settings like. preferences in other softwares.
> with tabs for each category and all?

No plans from me.  But, could potentially be added later.

**Reply by onegladiator4u** [reply]
ok brother . God bless . Thanks

**Comment by onegladiator4u** [comment]
hey brother ,
Can you tell me how can i change the location of the Unique Bible ?
its in My Documents now. i want it it in other drive.  Will moving the folder affect updating or running it?

**Reply by otseng** [reply]
You should be able to do that.  Just move the entire directory using file explorer.

**Reply by onegladiator4u** [reply]
also i would like to add a link to a file or folder relative to the uba folder. is that possible ?
like using `../` for a file that is in parent or grandparent folder.

**Reply by otseng** [reply]
Where do you want to add the link?

**Reply by onegladiator4u** [reply]
oh sorry to point it. On notes or books we create from html/notes

**Reply by otseng** [reply]
You can, but it just might need to be updated if you move your files.

**Reply by onegladiator4u** [reply]
i want to link it to note that if i move both Uba and those files together to another location it should work.
for example .
my App is in` F:\ `drive as 
![image](https://user-images.githubusercontent.com/48617390/125106355-2c544500-e0fd-11eb-9bcc-8edf07a29220.png)
i will be linking images and documents to the parent directory folder named` Bible Study`.
If i move the whole directory `Bible Study` shown above to` E:\ `drive it should work. How can i do that.?

**Reply by otseng** [reply]
> If i move the whole directory Bible Study shown above toE:\drive it should work. 

Yes, I think it should work.  

> How can i do that.?

You mean how to move a folder?  You should be able to just drag and drop it into F: drive.

**Reply by onegladiator4u** [reply]
i mean in note links .how can i links to folders, pdfs, videos , files etc..
![image](https://user-images.githubusercontent.com/48617390/125155342-bf808f80-e17c-11eb-914d-667c39405eff.png)
i want those to be clickable links. now in note editor there's no option to link folder or files. only hyperlinks and linking images

**Reply by onegladiator4u** [reply]
for the directory inside grandparent  for the directory `MP3's`
i tried backslash for directory names with quotes but it loses it's link click ability . it just shows plain text for these commands . i tested with
`***[CMD:::start ..\\../MP3/'s@OPEN directory on PC]`
`***[CMD:::start ..\\../MP3\'s@OPEN directory on PC]`
`***[CMD:::start ..\\../MP3'\s@OPEN directory on PC]`
`***[CMD:::start ..\\../MP3'/s@OPEN directory on PC]`
`***[CMD:::start ..\\../MP3'\\s@OPEN directory on PC]`
`***[CMD:::start ..\\../MP3'//s@OPEN directory on PC]`
`***[CMD:::start ..\\../MP3\\'s@OPEN directory on PC]`
`***[CMD:::start ..\\../MP3//'s@OPEN directory on PC]`

it shows as plain text 
![image](https://user-images.githubusercontent.com/48617390/125283569-2d23eb80-e336-11eb-9068-acb6e9833d5b.png)

**Reply by otseng** [reply]
> i tried backslash for directory names with quotes but it loses it's link click ability . it just shows plain text for these commands . i tested with
> `***[CMD:::start ..\\../MP3/'s@OPEN directory on PC]`

Try `***[CMD:::start ..\\..\\MP3@OPEN MP3 directory]`

All directory slashes should be `\\`

**Reply by onegladiator4u** [reply]
> Try `***[CMD:::start ..\\..\\MP3@OPEN MP3 directory]`

this works for `MP3` name. it won't work with `MP3's` . Folder name has quote. backslash  `MP3\'s`didn't work.

**Reply by otseng** [reply]
> this works for `MP3` name. it won't work with `MP3's` . Folder name has quote. backslash `MP3\'s`didn't work.

A quote (') in a directory or filename is not supported.  Any of these characters will cause it to fail:  '"*@[]

**Reply by onegladiator4u** [reply]
So I have the option to change the folder name only right?

**Reply by onegladiator4u** [reply]
i was actually planning on creating a reference book from the notes. i couldn't customize note like wanted to.
now i can customize and design that in html. i think doing the same type of linking in html is different right?
anchor tags in html doesn't work as normal inside the app.
how can i achieve the same effect of  `***[CMD:::start ..\\..\\MP3@OPEN MP3 directory]` in html that converts to book.

same code inside span look like this
![image](https://user-images.githubusercontent.com/48617390/125417485-e8c0ea4a-0775-43ad-8b53-d96a31a65b84.png)
i tried anchor tag. it just uses as inside search.

**Reply by otseng** [reply]
> So I have the option to change the folder name only right?

Correct, you cannot have a ' in a folder name in a notes file command.

> now i can customize and design that in html.

See:
https://github.com/eliranwong/UniqueBible/wiki/How-to-execute-a-command-in-a-book-generated-from-HTML

**Reply by onegladiator4u** [reply]
Thank You . This is working

**Comment by onegladiator4u** [comment]
Hey @otseng , is there a way to go to the titles sections in books. for eg:
we can go to book using `***[BOOK:::Josephus:::12 - Antiquities - Book10@Antiquities of the Jews - Book X Ch 9 Sec 7]`
it'll go to book 10. is there a way to got chapter 9 section 3?

**Reply by otseng** [reply]
There is no anchor tag for the chapters, so it's not possible.

**Reply by onegladiator4u** [reply]
oh ok.

**Reply by onegladiator4u** [reply]
do you know greek @otseng ?

**Reply by otseng** [reply]
No, but trying to learn it.

**Reply by onegladiator4u** [reply]
ok. i saw one scholar teach about Matthew 1:16 . In that 2 words `ἐγέννησεν` `ἐγεννήθη` one is mean `to beget` other is `was born`.
 That person taught that 'was born' is feminine and it can see in morphology or something . How can i see it in UBA. i checked it in app , i couldn't find much difference. Don't know whether i checked correctly .
i can see both point to  G1164; but they are different words.

**Reply by otseng** [reply]
Use Eliran's MIB Bible:
`BIBLE:::MIB:::Matthew 1:16`
Click on `V-AAI-3S`
Difference is first is active and second is passive.
Do you mean G1080?

**Reply by onegladiator4u** [reply]
> Do you mean G1080?

Yes. 

> Difference is first is active and second is passive.

i have to see what does that mean. Thank you @otseng

**Reply by onegladiator4u** [reply]
> There is no anchor tag for the chapters, so it's not possible.

@otseng what if the chapters have id's added manually ? how to navigate to an id?

**Reply by otseng** [reply]
> > There is no anchor tag for the chapters, so it's not possible.
> 
> @otseng what if the chapters have id's added manually ? how to navigate to an id?

Created issue https://github.com/eliranwong/UniqueBible/issues/755

**Reply by onegladiator4u** [reply]
Thank you @otseng

**Comment by onegladiator4u** [comment]
Hi brother ,
I downloaded Dakes commentary through app in Github commentaries.
It's not showing verses as link. what should I change to get it as other commentaries ?
![image](https://user-images.githubusercontent.com/48617390/126965608-050afb44-3ae8-4354-ac9f-4d814cd725b9.png)

**Reply by otseng** [reply]
> I downloaded Dakes commentary through app in Github commentaries.
> It's not showing verses as link. what should I change to get it as other commentaries ?

Yes, you're right.  It's missing the links in the file.  Can you create an issue for that?

**Reply by onegladiator4u** [reply]
ok.

**Comment by darrelwright** [comment]
You have to set parse on. Here's a link to my cloud that contains all of 
the Dake files. Be blessed!

https://u.pcloud.link/publink/show?code=kZnwOPXZ4pi1RveHToSYtOwluNfVnXG102Ky

Darrel

**Reply by onegladiator4u** [reply]
> https://u.pcloud.link/publink/show?code=kZnwOPXZ4pi1RveHToSYtOwluNfVnXG102Ky

@darrelwright Thanks brother for the link. i'll check it out.
could you help me which flag is that turns on? i have settings like this now.
![image](https://user-images.githubusercontent.com/48617390/126978704-ab116892-59e3-4ffd-9324-d2b6de7857f2.png)

**Reply by onegladiator4u** [reply]
@otseng @darrelwright i had problems with owenhebrew commentary too . this is different problem right ?
![image](https://user-images.githubusercontent.com/48617390/126984215-55a91df8-ea62-4b11-8995-8d47d1a4a6dc.png)

**Comment by onegladiator4u** [comment]
Hey @otseng ,
some commentary books don't have commentary on some books/chapters of the Bible. Is there a way to see what all the books/chapers that commentary have?
for example : i downloaded an old testament commentary named `Jewish Encyclopedia 1901-1906 `. when i choose a bible verse it'll show `['JE-TNK' does not contain this chapter.]`. how to know which book/chapter/verse have commentaries?

**Reply by otseng** [reply]
No easy way to do that now.  But that's a good idea to have a command/menu item to show what are the available commentaries for a particular chapter.  

Created issue for this:
https://github.com/eliranwong/UniqueBible/issues/711

**Reply by onegladiator4u** [reply]
great . thanks for looking in to it.

**Comment by onegladiator4u** [comment]
Hey @otseng can you help with somethings
1. if we wrire some references like` John 3:16` then after that how can i refer a verse like `18 `or` 19 `from same chapter 3 without writing `John 3:` in note so that it creates a hover link to that verse. 
![image](https://user-images.githubusercontent.com/48617390/128227045-520bf8cf-6c9f-4db3-b5cb-302f2e8d3d34.png)
2. How can i point a hover link to show LXX verse in bottom window ? is there a way.
![image](https://user-images.githubusercontent.com/48617390/128227631-547b8387-44ba-4a2c-a036-85b8431fe0ad.png)
i know that `***[BIBLE:::LXXUK:::Amos 7:1@Amos  7:1]` will take it to LXXUK. 
i just want to show in bottom window on hover.

**Reply by otseng** [reply]
> 1. if we wrire some references like` John 3:16` then after that how can i refer a verse like `18 `or`19`from same chapter 3 without writing `John 3:` in note so that it creates a hover link to that verse.

```
1. The people will be brought back home v. ***[BIBLE:::Ezekiel 37:21-22@21-22]
```

> 2. How can i point a hover link to show LXX verse in bottom window ? is there a way.

I don't think there's a way to specify a different Bible than what is being viewed in the Bible window.    Create an issue and I can add this.

**Reply by onegladiator4u** [reply]
> 1. The people will be brought back home v. ***[BIBLE:::Ezekiel 37:21-22@21-22]

lol. that was simple. Thank You 😊

> I don't think there's a way to specify a different Bible than what is being viewed in the Bible window. Create an issue and I can add this.

ok

**Comment by onegladiator4u** [comment]
hey @otseng ,
i was building a book with references to videos , images and documents in html. To convert to Book
if i use a ref tag to link a video file like
`<ref
                onclick="document.title='CMD:::start ..\\..\\Videos\\Learn\\Book_by_book\\Unlocking_the_New_Testament_by_David_Pawson\\Part_1_-_Matthew_1.mp4' ">
                Part 1</ref>`
It'll OPEN the Video file in external player (i prefer to use it than VLC).
But if i link a pdf file like 
`<ref
                onclick="document.title='CMD:::start ..\\..\\DOCUMENTS\\Bible_Books_Resources\\Thirdmill\\An-Introduction.pdf' ">
                An Introduction to Gospels</ref>`
 , it WONT OPEN in external reader(i prefer).
if i remove `\\An-Introduction.pdf`. it'll open the folder `Thirdmill` without a problem.
if i link to a png image like `3.png`, it'll OPEN.
does any default operations going on with a` .pdf` file?.
what should I do to open it to work as a book?

**Reply by otseng** [reply]
> what should I do to open it to work as a book?

I'll look into it.  Created issue https://github.com/eliranwong/UniqueBible/issues/749

**Reply by onegladiator4u** [reply]
The question should be ,what should I do to open it to work` in` a book?

**Reply by otseng** [reply]
> The question should be ,what should I do to open it to work` in` a book?

Created issue https://github.com/eliranwong/UniqueBible/issues/751

**Comment by onegladiator4u** [comment]
> Do you have a question on how to use a feature of UBA? Post it here.

Hi, what does the new share button do? nothing happened when I click on it.

**Comment by eliranwong** [comment]
After you click the share button, you should see a shared link ready for being copied and a qr code.

<img width="961" alt="Screenshot 2022-05-04 at 17 31 04" src="https://user-images.githubusercontent.com/25262722/166728490-2a4308cf-6a95-465a-870d-ba9605c7bcd5.png">

**Reply by eliranwong** [reply]
please read this wiki page: https://github.com/eliranwong/UniqueBible/wiki/Share-Button

**Reply by onegladiator4u** [reply]
> please read this wiki page: https://github.com/eliranwong/UniqueBible/wiki/Share-Button

This was helpful. I didn't have pillow installed . saw this on terminal error .

Does qr code open in full screen? i couldn't find close button. when i copied code it went background , then i closed via taskbar

**Reply by eliranwong** [reply]
The full screen issue should be fixed last night.  I uploaded the latest code to make the QR code open in a normal window rather than full screen

**Comment by onegladiator4u** [comment]
> Do you have a question on how to use a feature of UBA? Post it here.
Hey bro,
Now when we hover over a chapter bottom window shows like this 'Chapter 1'
![image](https://user-images.githubusercontent.com/48617390/188206456-281bb778-3435-4b05-a17f-fed76c7984a0.png)

is that information taking from a file or auto generated?
How can i add more information with Chapter 1, is it possible ?

**Reply by otseng** [reply]
It's hardcoded right now so there's no way to change it.  What would you like it to say?

**Reply by onegladiator4u** [reply]
Oh ok. No Problem. I was thinking of adding information for Psalm chapters like 
Genre: 
Attribute :
Book:
Themes:

**Reply by otseng** [reply]
> Oh ok. No Problem. I was thinking of adding information for Psalm chapters like Genre: Attribute : Book: Themes:

That'd be an interesting feature.  Created https://github.com/eliranwong/UniqueBible/issues/1197

**Reply by onegladiator4u** [reply]
That's great to hear bro. Thanks
